### PR TITLE
feat: train profile auto selection

### LINKED ISSUE #2155: [Phase 1 Profiles] HW auto-detect for CLI profiles
It would be nice to implement a way to auto-detect hardware, and suggest an appropriate default config profile for each of the components based on the hardware we detect. This will make for a smoother user experience that requires less work on the user's part to maximize the usage of their hardware and optimize performance (and get better output results.)

 
Feature Overview (aka. Goal Summary)
As a user, I want RHEL AI to detect the hardware I am working with so I do not have to choose/edit config files. 

Goals (aka. expected user outcomes)
This will make getting started with the process easier. Users will not have to look for relevant config files and reduces scope for errors.

Questions to Answer:
1. Will this be applicable for all hardware configurations (eg. RHEL AI suitable ones - 4xH100s and so on, as well as others, eg. 4xA10s)?
2. Will all the different components (eval, training, generate) be updated? 

Issues in this Epic:
- [ ] #2159

### PR DESCRIPTION
**Pull request by cdoern**
add auto selection for NVIDIA GPU train profiles. This uses torch under the hood
to read the amount of GPUs, tally the total vram, the GPU type, and the GPU amount. Use this information to choose the train profile which best matches the amount of GPUs/the GPU type. If none of these match, then match based off of vRAM. for good measure, set nproc_per_node at the end if basing of off vRAM. in case we found a profile which matches our vRAM amount but not our GPU amount

the workflow is as follows:

```ilab config init
Found /home/ec2-user/.config/instructlab/config.yaml, do you still want to continue? [y/N]: y
Existing training profiles were found in /home/ec2-user/.local/share/instructlab/internal/train_configuration/profiles
Do you also want to restore these profiles to the default values? [y/N]: 
Welcome to InstructLab CLI. This guide will help you to setup your environment.
Please provide the following values to initiate the environment [press Enter for defaults]:
Path to taxonomy repo [/home/ec2-user/.local/share/instructlab/taxonomy]: 
Path to your model [/home/ec2-user/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf]: 
Generating `/home/ec2-user/.config/instructlab/config.yaml`...
Detecting Hardware...
We chose NVIDIA 2x A100 as your designated training profile. This is for systems with 160 GB of vRAM.
Is this correct? [y/N]: y
```

TLDR: we choose a profile for you by default if we detect you have Nvidia hardware. If you do not like the one we choose for you, we prompt you to run through an interactive menu to choose your own.

resolves #2155 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/configuration.py [hunk 1]
```diff
@@ -1336,3 +1486,32 @@ def storage_dirs_exist() -> bool:
...
     return all(os.path.exists(dirpath) for dirpath in dirs_to_check)
+
+
+# render_configs_and_profiles takes the amount of GPUs on the system and returns a list of the following format:
+# [(MAX_VRAM_FOR_THIS_GROUP, [{"TRAIN_PROFILE_NAME": (VRAM_MAX_FOR_PROFILE, TRAIN_PROFILE)}....])]
+def render_profiles() -> list[tuple[int, list[dict[str, tuple[int, _train]]]]]:
+    return [
+        (
+            16,
+            [{"NVIDIA RTX 3070/3080/4070/4080": (16, SINGLE_CONSUMER_GPU_TRAIN)}],
```
Comment (jpodivin) [inline src/instructlab/configuration.py:1497]: What about the 4090 model, it has 24 GB and presumably would be something a high end desktop would have.

File: requirements.txt [hunk 2]
```diff
@@ -41,3 +41,5 @@ transformers>=4.40.0 ; python_version == '3.10'
 transformers>=4.41.2 ; python_version >= '3.11'
 trl>=0.9.4
 wandb>=0.16.4
+nvidia-ml-py==12.555.43
+nvidia_smi==0.1.3
```
Comment (jpodivin) [inline requirements.txt:45]: Why these libs?
nvidia_smi is not maintained very well, doesn't seem to have a link to source.
The nvidia-ml-py is better, but still, the same information could be obtained with torch, which is already in deps.

And as a bonus, it would allow for integration with other providers of high end ML HW. Like TPUs from Google.

```
for i in range(torch.cuda.device_count()):
  torch.cuda.get_device_properties(i)
_CudaDeviceProperties(name='Tesla T4', major=7, minor=5, total_memory=14915MB, multi_processor_count=40)
```

https://pytorch.org/docs/stable/cuda.html#module-torch.cuda

File: src/instructlab/configuration.py [hunk 3]
```diff
@@ -684,6 +684,313 @@ class Config(BaseModel):
     )
 
 
+EIGHT_GPU_TRAIN_AH = _train(
```
Comment (jpodivin) [inline src/instructlab/configuration.py:687]: Lot of these parameters are repeated among different configurations. I would recommend making them defaults, or derive a new class from train.

**Review by jpodivin — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 2 above)
Comment (cdoern) [reply requirements.txt]: thanks! switched to this

File: src/instructlab/configuration.py (hunk 3 above)
Comment (cdoern) [reply src/instructlab/configuration.py:682]: Internally its super helpful for us to have separate train profiles, eventually we want to derive a new class but that is out of scope for this PR. These all existed, I just moved them around

File: src/instructlab/config/init.py [hunk 4]
```diff
@@ -183,6 +195,48 @@ def init(
...
+        if vram < profile[0]:
+            for profile_names_and_values in profile[1]:
+                if train is not None:
+                    break
+                for i, vram_and_profile in enumerate(
+                    list(profile_names_and_values.values())
+                ):
+                    if vram < vram_and_profile[0]:
+                        chosen_profile_name = list(profile_names_and_values.keys())[i]
+                        chosen_vram = vram_and_profile[0]
```
Comment (danmcp) [inline src/instructlab/config/init.py:223]: Is this format standardized in some way or can we use a dict with named keys here?  It feels a bit cryptic to follow.

**Review by danmcp — CHANGES_REQUESTED** [review]

File: tests/testdata/leanimports.py [hunk 5]
```diff
@@ -5,7 +5,7 @@
 import sys
 
 # block slow imports
-for unwanted in ["deepspeed", "llama_cpp", "torch", "vllm"]:
+for unwanted in ["deepspeed", "llama_cpp", "vllm"]:
```
[the remark this answers is not in the document]
Comment (cdoern) [reply tests/testdata/leanimports.py:8]: sure, I can do it that way

File: src/instructlab/config/init.py [hunk 6]
```diff
@@ -183,6 +195,48 @@ def init(
...
+            properties = torch.cuda.get_device_properties(i)
+            total_vram += properties.total_memory  # memory in B
+    vram = convert_bytes_to_proper_mag(total_vram)[0]
+    for profile in profiles:
+        if vram < profile[0]:
+            for profile_names_and_values in profile[1]:
+                if train is not None:
+                    break
+                for i, vram_and_profile in enumerate(
+                    list(profile_names_and_values.values())
```
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/config/init.py:219]: I'll add a bunch of comments here, I realize its kind of confusing!!

File: src/instructlab/config/init.py [hunk 7]
```diff
@@ -117,53 +120,61 @@ def init(
...
-            click.echo(f"You selected: {profile_options[train_profile_selection - 1]}")
-            cfg.train = read_train_profile(
-                join(
-                    DEFAULTS.TRAIN_PROFILE_DIR,
-                    train_profile_filenames[train_profile_selection - 1],
-                )
+        chosen_profile, previous_vram, chosen_vram, cfg.train = hw_auto_detect()
+        yn = None
+        if chosen_vram is not None and chosen_profile is not None:
+            print(
```
Comment (nathan-weinberg) [inline src/instructlab/config/init.py:126]: Replace with logger.info

File: src/instructlab/config/init.py [hunk 8]
```diff
@@ -117,53 +120,61 @@ def init(
...
+                    )
             click.echo(
-                "No profile selected - any hardware acceleration for training must be configured manually."
+                "Please choose a train profile to use.\nTrain profiles assist with the complexity of configuring specific GPU hardware with the InstructLab Training library.\nYou can still take advantage of hardware acceleration for training even if your hardware is not listed."
             )
-        else:
-            click.secho(
-                "Invalid selection. Please select a valid train profile option.",
-                fg="red",
+            click.echo("[0] Choose for me (Hardware Defaults)")
```
Comment (nathan-weinberg) [inline src/instructlab/config/init.py:152]:
```suggestion
            click.echo("[0] Choose for me (Autodetect profile from system)")
```

Comment (nathan-weinberg) [reply src/instructlab/config/init.py:152]: Actually, taking a closer look at the logic, isn't this only reached if a user has already said "no" to the autodetected profile?

**Comment by nathan-weinberg** [comment]
Right now the user flow seems to be as follows (please correct me if I am wrong)

1. init
2. detect hardware
3. prompt to use autodetect or no
4. show menu if no

But it seems the menu now gives the option to do it again? I think the original "no profile" option should be there instead

an alternative user flow could be simply
1. init
2. show menu (options are [0] no profile [1] autodetect [2...n] manual profile option)

i think this might be more intuitive and save `ilab` some compute cycles from running autodetection when a user doesn't want it

**Comment by cdoern** [comment]
@nathan-weinberg your first point I agree with and will change but second I think having auto detection be the default behavior makes sense for most people.

**Comment by nathan-weinberg** [comment]
> @nathan-weinberg your first point I agree with and will change but second I think having auto detection be the default behavior makes sense for most people.

Chatted on Slack about this - will keep Charlie's flow but add additional detail to the user to see how their profile was modified if it was done so by the autodetection logic

File: src/instructlab/config/init.py [hunk 9]
```diff
@@ -183,6 +195,48 @@ def init(
...
+    for profile in profiles:
+        if vram < profile[0]:
+            for profile_names_and_values in profile[1]:
+                if train is not None:
+                    break
+                for i, vram_and_profile in enumerate(
+                    list(profile_names_and_values.values())
+                ):
+                    if vram < vram_and_profile[0]:
+                        chosen_profile_name = list(profile_names_and_values.keys())[i]
```
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/config/init.py:240]: switched, thanks!

File: src/instructlab/config/init.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/config/init.py:236]: added comments and switched to just using .items()

File: src/instructlab/config/init.py (hunk 4 above)
Comment (cdoern) [reply src/instructlab/config/init.py:241]: added comments why its this way, you can't put `_train` into a dict apparently, must be a tuple.

Comment (cdoern) [reply src/instructlab/config/init.py]: I could be wrong here, but that is why I did it initially bc python told me the type was un-mappable if I recall correctly

**Review by nathan-weinberg — COMMENTED** [review]
Thanks for addressing the comments Charlie, this is looking good to me - is it possible to add a unit or functional test for autodetection?

**Comment by cdoern** [comment]
@nathan-weinberg I will see if I can add a unit test. the issue is:

```
    # try nvidia
    if torch.cuda.is_available():
        click.echo("Detecting Hardware...")
        for i in range(torch.cuda.device_count()):
            properties = torch.cuda.get_device_properties(i)
            total_vram += properties.total_memory  # memory in B
    vram = convert_bytes_to_proper_mag(total_vram)[0]
```
but I am sure I can probably mock the value of total_vram somehow in the unit test?

File: src/instructlab/config/init.py (hunk 4 above)
Comment (danmcp) [reply src/instructlab/config/init.py]: Were you using _train as the key?  That wouldn't be allowed because it is mutable.  I am not sure of anything that would keep _train being used as a value.  Would be good to know the specific error if that was the case.

**Comment by cdoern** [comment]
@nathan-weinberg added a unit test

**Review by nathan-weinberg — APPROVED** [review]

**Comment by cdoern** [comment]
@nathan-weinberg I switched to click.echo as opposed to logger.info bc when using the logger the output looks bad as compared to click.echo (its used for everything else in `init`)

**Comment by nathan-weinberg** [comment]
@cdoern can you also update the CHANGELOG?

File: src/instructlab/config/init.py [hunk 10]
```diff
@@ -183,6 +202,60 @@ def init(
...
+    previous_vram = None
+    # try nvidia
+    if torch.cuda.is_available():
+        click.echo("Detecting Hardware...")
+        gpus = torch.cuda.device_count()
+        for i in range(gpus):
+            properties = torch.cuda.get_device_properties(i)
+            total_vram += properties.total_memory  # memory in B
+    vram = convert_bytes_to_proper_mag(total_vram)[0]
+    vram = 89
```
Comment (n1hility) [inline src/instructlab/config/init.py:224]: Looks like an unintentional double assign, debugging?

Comment (cdoern) [reply src/instructlab/config/init.py:224]: thanks Jason, totally missed this!!! Leftover debugging 😅

File: src/instructlab/config/init.py [hunk 11]
```diff
@@ -183,6 +202,59 @@ def init(
...
+        if vram < profile[0]:
+            # for each profile in this vRAM grouping of profiles
+            for profile_names_and_values in profile[1]:
+                if chosen_profile_name is not None:
+                    break
+                # enumerate over the profiles in the list, use .items() so we can access profile names
+                # the structure of this is due to the fact that _train instances cannot be members of a map. they must be in a tuple
+                for profile_name, profile_config in profile_names_and_values.items():
+                    # if our vram is < the vram required for this profile, choose it.
+                    assert isinstance(profile_config["vram"], int)
```
Comment (danmcp) [inline src/instructlab/config/init.py:236]: Suggestion: Rather than use asserts, if this is a condition we want to guard against, might want to check with an if and print a message and exit if not satisfied

File: src/instructlab/config/init.py [hunk 12]
```diff
@@ -183,6 +202,59 @@ def init(
...
+    chosen_profile_name = None
+    previous_vram = None
+    # try nvidia
+    if torch.cuda.is_available():
+        click.echo("Detecting Hardware...")
+        gpus = torch.cuda.device_count()
+        for i in range(gpus):
+            properties = torch.cuda.get_device_properties(i)
+            total_vram += properties.total_memory  # memory in B
+    vram = convert_bytes_to_proper_mag(total_vram)[0]
```
Comment (n1hility) [inline src/instructlab/config/init.py:223]: It's not a problem now but there is a risk of future bug here in that you are comparing the output of a routine that uses dynamic units with a routine that uses fixed units (render_profiles). This function convert_bytes_to_proper_mag() is a perfect printed output function, and could be extended to return TB, at which point a system with >= TB of vram would look like 1GB in the loop below.  There is also a less serious more of a nit issue that a float is being compared to an int.  I would recommend either just comparing in bytes, or if you want to downscale the unit convert it to an int of the floor of GB

File: src/instructlab/config/init.py [hunk 13]
```diff
@@ -183,6 +202,59 @@ def init(
...
+    for profile in profiles:
+        # for each profile mapping (vram -> list of profiles)
+        # if the vram we have is < the vram for this group of profiles
+        if vram < profile[0]:
```
Comment (n1hility) [inline src/instructlab/config/init.py:227]: Why is this organized as a group max and a separate profile max? Would it be simpler / more extensible to just have a lookup table for all known devices, with entries for # of GPUs? For example:
```
card_profiles = {
     'NVIDIA L40S', (2,  2xL40_PROFILE),
     'NVIDIA L40S', (4,  4xL40_PROFILE),
     ...
}
```

And then if you need a generic fallback you can change the device name to GENERIC_NVIDIA if the key is not in card_profiles.

Comment (cdoern) [reply src/instructlab/config/init.py:227]: I had the groupings almost for a future proofing purpose when we introduce system profiles. The top level "groups" would also have a `Config` Object associated with them for other cmd settings (generation serving etc). This might be a little too early for that though so I can change it to just a single map.


I think `GENERIC_NVIDIA` though will cause more issues than it'll solve for most users. given that for most usecases these profiles are about associating certain settings with a specific amount of vRAM. 

This way too we can give upstream users a better chance at matching their card.

Comment (cdoern) [reply src/instructlab/config/init.py]: lmk if what I changed it to is good for this first version of auto detection!

File: src/instructlab/config/init.py [hunk 14]
```diff
@@ -183,6 +210,52 @@ def init(
...
+    vram = int(floor(convert_bytes_to_proper_mag(total_vram)[0]))
+    if vram == 0:
+        return None, None, None, train, False
+    for profile, vram_and_train_config in profiles.items():
+        # if the vram we have is < the vram for profile
+        if not isinstance(vram_and_train_config["vram"], int) or not isinstance(
+            vram_and_train_config["train_config"], _train
+        ):
+            click.secho("Unable to retrieve train profiles", fg="red")
+            raise click.exceptions.Exit(1)
```
Comment (jpodivin) [inline src/instructlab/config/init.py:240]: You already know the exact data structure returned by `render_profiles`, checking it here is not necessary.

File: src/instructlab/config/init.py [hunk 15]
```diff
@@ -183,6 +210,52 @@ def init(
     )
 
 
+def hw_auto_detect() -> tuple[str | None, int | None, int | None, _train, bool]:
```
Comment (jpodivin) [inline src/instructlab/config/init.py:213]: The signature is pretty vague. When most items can be None you'll have to make a lot of checks later on.
At very least there should be a doc string, explaining exactly what they mean. A dataclass [1] for the result would be a better fit. Since the fields could have names then.

Overall, I would recommend rewriting this, so that returning `None` for most fields is not an option.

[1]https://docs.python.org/3/library/dataclasses.html#module-dataclasses

File: src/instructlab/configuration.py [hunk 16]
```diff
@@ -679,6 +679,313 @@ class Config(BaseModel):
...
+
+
+def finish_cfg_section(
+    defaults: dict, values: typing.Dict[str, typing.Any]
+) -> typing.Dict[str, typing.Any]:
+    for key, val in defaults.items():
+        if values.get(key) is None and val is not None:
+            values[key] = val
+
+    return values
```
Comment (jpodivin) [inline src/instructlab/configuration.py:986]: You don't seem to use this anywhere?

File: src/instructlab/config/init.py [hunk 17]
```diff
@@ -117,53 +121,76 @@ def init(
...
+            click.echo("[0] No profile (CPU, Apple Metal, AMD ROCm)")
+            for i, value in enumerate(profile_options):
+                click.echo(f"[{i+1}] {value}")
+            train_profile_selection = click.prompt(
+                "Enter the number of your choice [hit enter for hardware defaults]",
+                type=int,
+                default=0,
             )
-            raise click.exceptions.Exit(1)
+            if 1 <= train_profile_selection <= len(profile_options):
```
Comment (jpodivin) [inline src/instructlab/config/init.py:171]: `in` would work out as well, although you would need to decrement.

File: src/instructlab/config/init.py (hunk 14 above)
Comment (cdoern) [reply src/instructlab/config/init.py:240]: this is necessary for two reasons:
1. mypy demands vaidation of the types before use
2. for future code changes this will be a nice catch if anyone breaks the render_profiles funcion

File: src/instructlab/config/init.py [hunk 18]
```diff
@@ -183,6 +210,52 @@ def init(
...
+            train = vram_and_train_config["train_config"]
+            break
+        # keep track of the floor since its helpful to print the vram range
+        previous_vram = vram_and_train_config["vram"]
+    if train is not None:
+        edited_cfg = False
+        if train.nproc_per_node != gpus:
+            # edit nproc per node in the case that we didnt match an exact profile, so we use all GPUs on our system
+            train.nproc_per_node = gpus
+            edited_cfg = True
```
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/config/init.py:254]: a core tenet of hw auto detection is a best guess estimate for users who we dont have a direct profile for.

If we provide them with a profile that is pretty good, but has the wrong GPU number, then the profile is relatively useless. In this scenario we also log to warn the user this case has happened

[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/config/init.py:254]: Otherwise we will be in the business of creating profiles for all graphics cards

File: src/instructlab/config/init.py [hunk 19]
```diff
@@ -183,6 +210,52 @@ def init(
     )
 
 
+def hw_auto_detect() -> tuple[str | None, int | None, int | None, _train, bool]:
+    # Third Party
+    import torch
```
[the remark this answers is not in the document]
Comment (cdoern) [reply src/instructlab/config/init.py:215]: for our test suite and accorss ilab in general torch is imported inside of functions where it is needed, not globally.

File: src/instructlab/config/init.py (hunk 17 above)
Comment (cdoern) [reply src/instructlab/config/init.py:171]: this is existing behavior of the profile output. There also isn't a profile for [0] which is why this was added. Not looking to change functionality like this for no reason in this PR. It just looks like I changed it because the code moved

File: src/instructlab/config/init.py (hunk 12 above)
Comment (n1hility) [reply src/instructlab/config/init.py]: Cool. Yes, the recently released H200 systems are available for use/purchase have > 1TB (1128 GB x 8 - (141GB per card)). We definitely want to run on those.

File: src/instructlab/config/init.py (hunk 15 above)
Comment (cdoern) [reply src/instructlab/config/init.py:213]: for this first version of auto detection i'd like to opt to keep it this way, this is a high priority feature we want to get in today or tomorrow

File: src/instructlab/config/init.py (hunk 13 above)
Comment (n1hility) [reply src/instructlab/config/init.py]: Sorry my reply was a bit too vague. What I meant was that essentially you are implementing GENERIC_NVIDIA, in that you are using a heuristic (vram sizing) to determine which node, vs a known static association. What I meant is that a fixed association with a heuristic as a fallback would have two advantages:

- Simpler implementation, better code clarity
- Less side-effect potential

Using vram is a good choice for an input to a heuristic, but as any heuristic it's only mostly accurate. As an example, a 2x80 H100 and a 4x40 A100 both render the same vram, but we want two different configurations. In the former we need a gpu count of 2 for inferencing (I know this PR isnt addressing that so I am speaking broadly), and in the latter we want 4. We also have architectural generation differences so there may be training features which start to appear in training configurations (e.g. FP8 or other sm_89..sm_90 specific optos).

The advantage of a fixed association comes into play if we need to add a profile variation. Let's say we are close to final release and we need to tweak something just for a specific GPU (say a 4xH200), and so we add some new profiles. The logic will only execute differently only for those specific profiles since it's just a lookup table. In contrast, you might tweak the heuristic algorithm, and now the behavior potentially affects everything, so you could end up with a side-effect breaking another card type unintentionally, and then you update it again and potentially break another and so on. The only away around that is an extensive test that verifies all permutations, wherea a lookup table you can verify by proof and lighter testing.

There is a trade-off though in that you have to update it, for new hardware, but falling back to the heuristic + a generic safe config helps there. For example the generic nvidia configs could assume a low arch sm_70, and then just be sized by vram similar to how you have it. I am traveling today but will comment on your next iteration.

File: src/instructlab/config/init.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (jpodivin) [reply src/instructlab/config/init.py:254]: No. You could just fail, and let user define their own config. This way you are covering a misconfiguration. Instead you should allow for profiles to be loaded from a file, say `toml` format or `yaml`, doesn't really matter.
That way, when users have different set of GPUs, they can just add a file with their config and load that.

File: src/instructlab/config/init.py (hunk 13 above)
Comment (cdoern) [reply src/instructlab/config/init.py]: @n1hility how about something like a:

`--choose-closest` flag that if specified will match the closest vRAM, otherwise will match based off the GPU name? I feel like this matches both of the usecases we are worried about?

File: src/instructlab/config/init.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (jpodivin) [reply src/instructlab/config/init.py:215]: It's an anti pattern. It shouldn't be done as it brings very little benefit and makes it harder to see dependencies of the module.
https://peps.python.org/pep-0008/#imports

File: src/instructlab/config/init.py (hunk 15 above)
Comment (jpodivin) [reply src/instructlab/config/init.py:213]: Temporary solutions like these tend to stick around. At very least you need to provide information about the structure returned. There is no way to infer what those values mean, without looking at the code. Type hints are not enough.

File: src/instructlab/config/init.py (hunk 13 above)
Comment (danmcp) [reply src/instructlab/config/init.py]: > `--choose-closest` flag that if specified will match the closest vRAM, otherwise will match based off the GPU name? I feel like this matches both of the usecases we are worried about?

I might have this wrong but the discovery of such a flag seems unlikely.  Unless you are automating the init call and know that provides a good outcome for your hardware.

Otherwise I would think we have one shot to guess, followed by the user potentially deciding we made a made choice.  If that is the case, seems like we should just focus on making the best guess we can given the hardware.

Comment (cdoern) [reply src/instructlab/config/init.py]: >  If that is the case, seems like we should just focus on making the best guess we can given the hardware.

Is the best guess here based off of vRAM or card name?

The version I have now is more of a "best guess" but will also very clearly match the supported d/s hardware.

This new proposal would only match based off of GPU name and amount, no guessing involved.

Comment (cdoern) [reply src/instructlab/config/init.py]: How about this:

1. check for matching GPU name and Number
2. if none of these match, try to match by vRam

This accounts for
a. the upstream usecase where people just want it to work
b. the concern jason raised with using the wrong heuristic as our first check
c. makes sure all profiles we support get chosen before any type of smart "guessing:

Comment (nathan-weinberg) [reply src/instructlab/config/init.py]: I like the above idea of GPU card name followed by matching by vRAM - if neither match I say we fail out with some message like "couldn't autodetect - please select a profile manually"

Comment (n1hility) [reply src/instructlab/config/init.py]: @cdoern +1 having the heuristic fallback is important.

**Comment by cdoern** [comment]
@n1hility @danmcp this should be set based on the convo in the above thread about the proper heuristic. The workflow is now:

1. match based off of GPU name and number.
2. If none of those match return the profile best suited based off of vRAM.
3. if 2. is the case, tell the user this is a best approx and none of our profiles matched exactly

File: src/instructlab/config/init.py (hunk 15 above)
Comment (cdoern) [reply src/instructlab/config/init.py]: made an issue to track this as a follow up #2275

File: src/instructlab/config/init.py [hunk 20]
```diff
@@ -183,6 +209,70 @@ def init(
...
+        gpus = torch.cuda.device_count()
+        for i in range(gpus):
+            properties = torch.cuda.get_device_properties(i)
+            gpu_name = properties.name
+            total_vram += properties.total_memory  # memory in B
+    vram = int(floor(convert_bytes_to_proper_mag(total_vram)[0]))
+    if vram == 0:
+        return None, None, train, False
+    profiles = PROFILE_MAPPINGS
+    for group_gpu_name, list_of_count_vram_and_config in profiles.items():
```
Comment (n1hility) [inline src/instructlab/config/init.py:235]: I would recommend going with a lookup approach which should simplify the code further.  Two possible approaches:

Default Fallback Approach

```
def lookup_card(gpu_name, gpu_count, vram):
     card_entry = PROFILE_MAPPINGS.get(gpu_name, UNKOWN_NVIDIA)
     # select_card will examine the tuple/list for a matching gpu count and card vram entry (if variants)
     return select_best_card_config(card_entry, gpu_count, vram)
```

UNKNOWN_NVIDIA would point to a tuple/list that would contain the lowest common denominator on gpu capabilities with a number of entries on GPUs and sizes, this might be an alias to a specific config for this purpose or it could be an alias to an another key

Scan All Fallback

Alternatively could scan the table similar to this code but with the direct lookup and a tighter scan loop:

```
def lookup_card(gpu_name, gpu_count, vram):
     card_entry = PROFILE_MAPPINGS.get(gpu_name, UNKOWN_NVIDIA)
     # select_card will examine the tuple/list for a matching gpu count and card vram entry (if variants)
     if card_entry is not None:
          return select_best_card_config(card_entry, gpu_count, vram)

    return find_closest_gpu_vram_match(gpu_count, vram)
```

Comment (cdoern) [reply src/instructlab/config/init.py:235]: ok, thanks jason! I will implement one of the above

Comment (cdoern) [reply src/instructlab/config/init.py:235]: there is one issue with:
```
def lookup_card(gpu_name, gpu_count, vram):
     card_entry = PROFILE_MAPPINGS.get(gpu_name)
     # select_card will examine the tuple/list for a matching gpu count and card vram entry (if variants)
     if card_entry is not None:
          return select_best_card_config(card_entry, gpu_count, vram)

    return find_closest_gpu_vram_match(gpu_count, vram)
    ```
    it is not a sure bet that just because we match the GPU name we find a profile with vRAM and GPU count. So there is a scenario where we fall back on the vram based match

this is fine with me, but its basically the same end result as what I currently have

I am going to push a version with these abstractions soon though because it helps to have sep functions

Comment (cdoern) [reply src/instructlab/config/init.py]: switched to this structure, PTAL!

**Comment by danmcp** [comment]
@mergifyio rebase

File: src/instructlab/config/init.py [hunk 21]
```diff
@@ -183,6 +208,113 @@ def init(
...
+def lookup_card(
+    gpu_name, gpu_count, vram
+) -> tuple[bool, int | None, str | None, _train]:
+    profiles = PROFILE_MAPPINGS
+    card_entry = profiles.get(gpu_name)
+    chosen_profile_name = None
+    if card_entry is not None:
+        chosen_profile_name = f"Nvidia {gpu_count}x {gpu_name}"
+        # it is not guaranteed we match, even if we find an entry
+        # what if the vram isnt found or what if the gpu count doesn't match
```
Comment (danmcp) [inline src/instructlab/config/init.py:254]: nit: isn't

File: src/instructlab/config/init.py [hunk 22]
```diff
@@ -183,6 +208,113 @@ def init(
...
+        matched, train = match_prfofile_based_on_gpu_count(
```
Comment (danmcp) [inline src/instructlab/config/init.py:255]: Is prfofile a typo?

File: src/instructlab/config/init.py [hunk 23]
```diff
@@ -117,53 +121,74 @@ def init(
...
+                elif len(split_filename) == 3:
+                    gputype1 = split_filename[0]
+                    gputype2 = split_filename[1]
+                    num_gpu = split_filename[2].split(".")[0]
+                    profile_options.append(
+                        f"Nvidia {gputype1}/{gputype2} {num_gpu} ({filename})"
+                    )
             click.echo(
-                "No profile selected - any hardware acceleration for training must be configured manually."
+                "Please choose a train profile to use.\nTrain profiles assist with the complexity of configuring specific GPU hardware with the InstructLab Training library.\nYou can still take advantage of hardware acceleration for training even if your hardware is not listed."
```
Comment (danmcp) [inline src/instructlab/config/init.py:160]: Nit: ... complexity of configuring InstructLab training for specific GPU hardware.

File: src/instructlab/config/init.py [hunk 24]
```diff
@@ -183,6 +208,113 @@ def init(
...
+        click.echo("Detecting Hardware...")
+        gpus = torch.cuda.device_count()
+        for i in range(gpus):
+            properties = torch.cuda.get_device_properties(i)
+            gpu_name = properties.name
+            total_vram += properties.total_memory  # memory in B
+    vram = int(floor(convert_bytes_to_proper_mag(total_vram)[0]))
+    if vram == 0:
+        return None, None, train, False
+    not_exact_match, chosen_vram, chosen_profile_name, train = lookup_card(
```
Comment (danmcp) [inline src/instructlab/config/init.py:233]: This is confusing to read that we flip the value of matched inside of lookup_card when returned from the matching function.  Was there a good reason to flip the bool?

**Review by danmcp — CHANGES_REQUESTED** [review]
Few nits that are probably worth fixing.  Happy with where this has gotten to!

File: src/instructlab/config/init.py (hunk 22 above)
Comment (cdoern) [reply src/instructlab/config/init.py:255]: yes, sorry! fixing now

File: src/instructlab/config/init.py (hunk 24 above)
Comment (cdoern) [reply src/instructlab/config/init.py:233]: good point, this is confusing. I don't feel strongly about the way I have it now so I'll switch it to make more sense

**Review by danmcp — APPROVED** [review]

**Comment by cdoern** [comment]
waiting on a final lgtm from @n1hility before merging since he had some comments I took into account

File: src/instructlab/config/init.py [hunk 25]
```diff
@@ -183,6 +208,114 @@ def init(
...
+# match_profile_based_on_vram chooses a training profile based off the amount of vram in each profile
+# this profile will be modified to match the number of GPUs on the user's system
+def match_profile_based_on_vram(vram) -> tuple[int | None, str | None, _train]:
+    chosen_profile_name = None
+    chosen_vram = None
+    train = _train()
+    for group_gpu_name, list_of_count_vram_and_config in PROFILE_MAPPINGS.items():
+        for gpu_count_configs in list_of_count_vram_and_config:
+            if (
+                not isinstance(gpu_count_configs["gpu_count"], int)
```
Comment (n1hility) [inline src/instructlab/config/init.py:274]: I assume the isinstance is intended to check if it exists right? If so there will be a keyerror here when it doesnt exist which looks like the code block isnt expecting as an outcome.

**Comment by n1hility** [comment]
LGTM other than that one issue above

File: src/instructlab/config/init.py (hunk 25 above)
Comment (cdoern) [reply src/instructlab/config/init.py:274]: this is more for a mypy check since it didn't like us using the fields before checking the type.

I guess if someone messes with the struct this will error out, i can probably handle this more elegantly. The Pr merged because you were the last requested reviewer, but I will follow up with a PR to make this better.
